src/nn/visitors.rs

```diff
@@ -124,131 +123,51 @@ pub(crate) struct RecursiveWalker<'a, M, F> {
-            m: get_refs(self.m),
-            f: self.f,
-            path: self.path,
-        };
-        Field::iter_tensors(&mut walker)?;
-        self.path.pop();
-        Ok(())
-    }
-    fn visit_tensor<S: Shape, GetRef, GetMut>(
-        &mut self,
-        mut get_refs: GetRef,
-        _: GetMut,
-        name: &str,
-        opts: TensorOptions<S, E, D>,
-    ) -> Result<(), F::Err>
-    where
-        GetRef: FnMut(&M) -> &Tensor<S, E, D>,
-        GetMut: FnMut(&mut M) -> &mut Tensor<S, E, D>,
-    {
-        self.path.push(name.into());
-        self.f.visit(self.path.join("."), opts, get_refs(self.m))?;
-        self.path.pop();
-        Ok(())
-    }
-}
-
-impl<'a, M, E: Dtype, D: DeviceStorage, F: VisitTensorMut<E, D>> TensorVisitor<M, E, D>
-    for RecursiveWalker<'a, &'a mut M, F>
+impl<'a, M, E: Dtype, D: DeviceStorage, F: VisitTensors<E, D>> TensorVisitor<M, E, D>
+    for RecursiveWalker<'a, ContainerWithModule<'a, F::Container, M>, F>
 {
     type Err = F::Err;
-    fn visit_module<Field, GetRef, GetMut>(
-        &mut self,
-        _: GetRef,
-        mut get_muts: GetMut,
-        name: &str,
-    ) -> Result<(), F::Err>
-    where
-        GetRef: FnMut(&M) -> &Field,
-        GetMut: FnMut(&mut M) -> &mut Field,
-        Field: TensorCollection<E, D>,
-    {
-        self.path.push(name.into());
-        let mut walker = RecursiveWalker {
-            m: get_muts(self.m),
-            f: self.f,
-            path: self.path,
-        };
-        Field::iter_tensors(&mut walker)?;
-        self.path.pop();
-        Ok(())
-    }
-    fn visit_tensor<S: Shape, GetRef, GetMut>(
-        &mut self,
-        _: GetRef,
-        mut get_muts: GetMut,
-        name: &str,
-        opts: TensorOptions<S, E, D>,
-    ) -> Result<(), F::Err>
-    where
-        GetRef: FnMut(&M) -> &Tensor<S, E, D>,
-        GetMut: FnMut(&mut M) -> &mut Tensor<S, E, D>,
-    {
-        self.path.push(name.into());
-        self.f.visit(self.path.join("."), opts, get_muts(self.m))?;
-        self.path.pop();
-        Ok(())
-    }
-}
 
-impl<'a, M, E: Dtype, D: DeviceStorage, F: VisitTensorMutRef<E, D>> TensorVisitor<M, E, D>
-    for RecursiveWalker<'a, (&'a mut M, &'a M), F>
-{
-    type Err = F::Err;
     fn visit_module<Field, GetRef, GetMut>(
         &mut self,
         mut get_refs: GetRef,
         mut get_muts: GetMut,
         name: &str,
-    ) -> Result<(), F::Err>
+    ) -> Result<(), Self::Err>
     where
         GetRef: FnMut(&M) -> &Field,
         GetMut: FnMut(&mut M) -> &mut Field,
         Field: TensorCollection<E, D>,
     {
         self.path.push(name.into());
         let mut walker = RecursiveWalker {
-            m: (get_muts(self.m.0), get_refs(self.m.1)),
+            m: F::Container::get_field(&mut self.m, &mut get_refs, &mut get_muts),
             f: self.f,
             path: self.path,
         };
         Field::iter_tensors(&mut walker)?;
+        std::mem::drop(walker);
```

Comment:
Is this necessary?

Reply:
Yes, if you remove it, rust will consider `self.path` to be mutably borrowed on the next line.